### PR TITLE
COM-6142

### PR DESCRIPTION
**Pull request by mcesariniflu**
modified:   fluster/test_suite.py
	modified:   fluster/utils.py
	new file:   scripts/gen_av1_argon.py
	new file:   scripts/test_zip/SVCBC-1/SVCBC-1-L0.264
	new file:   scripts/test_zip/SVCBC-1/SVCBC-1-L1.264
	new file:   scripts/test_zip/SVCBC-1/SVCBC-1.264
	new file:   test_suites/av1/AV1_ARGON_VECTORS_MANUAL.json

### CODE REVIEW

**Comment by mdimopoulos** [comment]
Important time stats
- scripts/gen_av1_argon.py -> took about **30 minutes** to download, extract, read checksums and create the test suite
- ./fluster.py download AV1_ARGON_VECTORS -> after having created the test suite file from above step, downloading the source file and extracting the vectors took **2h and 47 minutes**

So in a future iteration we should try to bring the download logic of fluster close to the one used in the test suite generator

File: scripts/gen_av1_argon.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+                    try:
+                        result = utils.run_command_with_output(command).splitlines()
+                        pix_fmt = result[0]
+                        if pix_fmt == "unknown":
+                            pix_fmt = "Unknown"
+                    except subprocess.CalledProcessError:
+                        pix_fmt = "None"
+
+                # Processing md5 files
+                md5_file_to_find= os.path.splitext(filename)[0] + '.md5'
```
Comment (mdimopoulos) [inline scripts/gen_av1_argon.py:121]:
```suggestion
                md5_file_to_find = os.path.splitext(filename)[0] + '.md5'
```

File: scripts/gen_av1_argon.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+                split = full_path.split('/')
+                md5_directory = '/'.join(split[:9]) + '/' + 'md5_ref'
+                md5_path_to_file = os.path.join(md5_directory, md5_file_to_find)
+                
+                # Check the .md5 file and get checksum
+                if os.path.exists(md5_path_to_file):
+                    try:
+                        result_checksum = self._fill_checksum_argon(self,md5_path_to_file)
```
Comment (mdimopoulos) [inline scripts/gen_av1_argon.py:129]:
```suggestion
                        result_checksum = self._fill_checksum_argon(self, md5_path_to_file)
```

File: scripts/gen_av1_argon.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+                    except:
+                        raise Exception("MD5 does not match")
+                else:
+                    try:
+                        result_checksum = utils.file_checksum(full_path)
+                    except:
+                        raise Exception("MD5 cannot be calculated")
+
+                # Add data to the Test vector and the Test suite
```
Comment (mdimopoulos) [inline scripts/gen_av1_argon.py:138]:
```suggestion
                # Add data to the test vector and the test suite
```

File: scripts/gen_av1_argon.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+                md5_file_to_find= os.path.splitext(filename)[0] + '.md5'
+                split = full_path.split('/')
+                md5_directory = '/'.join(split[:9]) + '/' + 'md5_ref'
+                md5_path_to_file = os.path.join(md5_directory, md5_file_to_find)
+                
+                # Check the .md5 file and get checksum
+                if os.path.exists(md5_path_to_file):
+                    try:
+                        result_checksum = self._fill_checksum_argon(self,md5_path_to_file)
+                    except:
```
Comment (mdimopoulos) [inline scripts/gen_av1_argon.py:130]:
```suggestion
                    except Exception as ex:
```

File: scripts/gen_av1_argon.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+                        raise Exception("MD5 does not match")
+                else:
+                    try:
+                        result_checksum = utils.file_checksum(full_path)
+                    except:
```
Comment (mdimopoulos) [inline scripts/gen_av1_argon.py:135]:
```suggestion
                    except Exception as ex:
```

File: scripts/gen_av1_argon.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        with open(checksum_file, "r") as checksum_file:
+            regex = re.compile(r"([a-fA-F0-9]{32,}).*\.(yuv|rgb|gbr)")
+            lines = checksum_file.readlines()
+            if any((match := regex.match(line)) for line in lines):
+                result = match.group(1)[:32].lower()
+            else:
+                result = -1
+            # Assert that we have extracted a valid MD5 from the file
+            # assert len(test_vector.result) == 32 and re.search(
+            #     r"^[a-fA-F0-9]{32}$", test_vector.result) is not None, f"{test_vector.result} is not a valid MD5 hash"
```
Comment (mdimopoulos) [inline scripts/gen_av1_argon.py:159]: This seems obsolete

File: fluster/utils.py [hunk 7]
```diff
@@ -124,6 +124,26 @@ def extract(filepath: str, output_dir: str, file: Optional[str] = None) -> None:
     else:
         raise Exception(f"Unknown tarball format {filepath}")
 
+def copy(origin_path, dest_path, filename):
```
Comment (mdimopoulos) [inline fluster/utils.py:127]: I don't see this function being used anywhere. Can you double-check?

**Review by mdimopoulos — CHANGES_REQUESTED** [review]

File: fluster/test_suite.py [hunk 8]
```diff
@@ -224,6 +225,54 @@ def _download_worker(ctx: DownloadWork) -> None:
             if not ctx.keep_file:
                 os.remove(dest_path)
 
+    @staticmethod
+    def _download_worker_av1_argon(ctx: DownloadWork) -> None:
```
[the remark this answers is not in the document]
Comment (mdimopoulos) [reply fluster/test_suite.py:229]: This suggestion should be implemented in a future task, not in this one.

File: fluster/test_suite.py [hunk 9]
```diff
@@ -255,20 +320,33 @@ def _callback_error(err: Any) -> None:
...
+                            error_callback=_callback_error,
+                        )
+                    )
+                else:
+                    downloads.append(
+                        pool.apply_async(
+                            self._download_worker,
+                            args=(dwork,),
+                            error_callback=_callback_error,
+                        )
```
Comment (rgonzalezfluendo) [inline fluster/test_suite.py:337]: nitpick:

```suggestion
                if self.name == "AV1_ARGON_VECTORS":
                    download_worker = self._download_worker_av1_argon
                else:
                    download_worker = self._download_worker
                downloads.append(
                    pool.apply_async(
                        download_worker,
                        args=(dwork,),
                        error_callback=_callback_error,
                    )
                )
```

File: fluster/test_suite.py [hunk 10]
```diff
@@ -255,20 +320,33 @@ def _callback_error(err: Any) -> None:
...
             pool.close()
             pool.join()
 
         for job in downloads:
             if not job.successful():
                 sys.exit("Some download failed")
 
+        if self.name == "AV1_ARGON_VECTORS":
+            if not dwork.keep_file:
+                os.remove(os.path.join(dwork.out_dir,dwork.test_suite_name, os.path.basename(test_vector.source)))
```
[the remark this answers is not in the document]
Comment (mcesariniflu) [reply fluster/test_suite.py:348]: The download worker has the value in this context so I changed the variable test_vector.source for dwork.test_vector.source.

File: scripts/gen_av1_argon.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+                exception_str = ""
+                utils.download(self.site + self.name, extract_folder)
+            except urllib.error.URLError as ex:
+                exception_str = str(ex)
+                print(f"\tUnable to download {self.site + self.name} to {extract_folder}, {exception_str}")
+            except Exception as ex:
+                raise Exception(str(ex)) from ex
+
+        # Unzip the file
+        source_url = self.site + self.name
```
[the remark this answers is not in the document]
Comment (mcesariniflu) [reply scripts/gen_av1_argon.py:83]: I did it, and I have modified the lines 72, 75, and 78.

File: fluster/utils.py (hunk 7 above)
Comment (mcesariniflu) [reply fluster/utils.py:127]: I have removed it.

**Review by mdimopoulos — COMMENTED** [review]
All pending points have been addressed 👍🏼.

I encountered issues running `gen_av1_argon.py` successfully on MacOS. I decided to push a fixup commit that solves any problems.
@mcesariniflu let me know if the latest version of the script runs well on Ubuntu.

**Review by mdimopoulos — APPROVED** [review]
I tried that it works fine on Ubuntu 20. Approved

**Pull request merged**